de.peeeq.wurstscript/src/main/java/de/peeeq/wurstscript/intermediatelang/optimizer/BranchMerger.java

```diff
@@ -0,0 +1,109 @@
+package de.peeeq.wurstscript.intermediatelang.optimizer;
+
+import com.google.common.collect.Lists;
+import de.peeeq.wurstscript.intermediatelang.optimizer.ControlFlowGraph.Node;
+import de.peeeq.wurstscript.jassIm.*;
+import de.peeeq.wurstscript.translation.imtranslation.ImTranslator;
+import org.eclipse.jdt.annotation.Nullable;
+
+import java.util.List;
+
+/**
+ * merges identical nodes in branches if possible without side effects
+ * <p>
+ * the input must be a flattened program
+ */
+public class BranchMerger {
+    public int totalLocalsMerged = 0;
+    private final ImProg prog;
+    private final ImTranslator trans;
+
+    public BranchMerger(ImTranslator trans) {
+        this.prog = trans.getImProg();
+        this.trans = trans;
+    }
+
+    public void optimize() {
+        for (ImFunction func : prog.getFunctions()) {
+            optimizeFunc(func);
+        }
+    }
+
+    private void optimizeFunc(ImFunction func) {
+        mergeBranches(func);
+    }
+
+
+    private void mergeBranches(ImFunction func) {
+        ControlFlowGraph cfg = new ControlFlowGraph(func.getBody());
+
+        // init in and out with empty sets
+        for (Node n : cfg.getNodes()) {
+            @Nullable ImStmt stmt = n.getStmt();
+            if (stmt != null && stmt.getParent() != null && stmt.getParent() instanceof ImIf) {
+                ImIf imIf = (ImIf) stmt.getParent();
+                if (n.getPredecessors().size() <= 1 && n.getSuccessors().size() == 2) {
+                    Node leftStmt = n.getSuccessors().get(0);
+                    Node rightStmt = n.getSuccessors().get(1);
+                    if (leftStmt.toString().equals(rightStmt.toString())) {
+                        if (n.getPredecessors().size() == 1) {
+                            // Possible match. At last check if condition causes sideeffects.
+                            if (hasNoSideEffects(n, leftStmt)) {
+
+                                ImStmt mergedStmt = leftStmt.getStmt().copy();
```

Comment:
Not really necessary to copy here. You can probably just remove it from the old if-statement and add it to the new position.